torch/distributed/device_mesh.py

```diff
@@ -326,11 +326,10 @@ def __hash__(self):
         def __eq__(self, other: object) -> bool:
             if not isinstance(other, DeviceMesh):
                 return False
-            if id(self.mesh) == id(other.mesh):
-                return True
             return (
                 self.mesh.shape == other.mesh.shape
                 and self._flatten_mesh_list == other._flatten_mesh_list
+                and id(self.mesh) == id(other.mesh)
```

Comment:
If we are already hashing `id()`, do we need to hash the others? It seems like we could not override `__hash__` and `__eq__` and instead use the default one, which I think uses `id()`.

If we want to hash `mesh.shape` and `_flatten_mesh_list` still, then a nit could be to move `id(self.mesh) == id(other.mesh)` to be the 1st check since that is the most likely to fail.

Reply:
Makes sense. Thanks for reviewing. I agree that if we are already hashing `id()`, then we don't need to hash the others (`mesh.shape` and `_flatten_mesh_list`). Thanks for the suggestion. 

@wanchaol Want to run by you since we are changing the behavior of `hash` and `__eq__`.